cnd/src/ethereum.rs

```diff
@@ -1,22 +1,140 @@
 #![warn(unused_extern_crates, missing_debug_implementations, rust_2018_idioms)]
 #![forbid(unsafe_code)]
 
+use bitcoin::hashes::core::{fmt::Formatter, str::FromStr};
+use core::fmt;
 pub use ethbloom::{Bloom as H2048, Input};
-pub use primitive_types::{H160, H256, U128, U256};
+use hex::FromHexError;
+pub use primitive_types::U256;
 use serde::{Deserialize, Serialize};
 use serde_hex::{CompactPfx, SerHex, SerHexSeq, StrictPfx};
-
-pub type Address = H160;
+use std::fmt::Display;
 
 #[derive(Debug, Default, Copy, Clone, PartialEq, Deserialize)]
 pub struct H64(#[serde(with = "SerHex::<StrictPfx>")] [u8; 8]);
 
+#[derive(
+    Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
+)]
+pub struct Address(#[serde(with = "SerHex::<StrictPfx>")] [u8; 20]);
+
+impl Address {
+    pub fn from_slice(src: &[u8]) -> Self {
+        let mut address = Address([0u8; 20]);
+        address.0.copy_from_slice(src);
+        address
+    }
+}
+
+impl From<[u8; 20]> for Address {
+    fn from(bytes: [u8; 20]) -> Self {
+        Address(bytes)
+    }
+}
+
+impl From<Address> for [u8; 20] {
+    fn from(s: Address) -> Self {
+        s.0
+    }
+}
+
+impl FromStr for Address {
+    type Err = FromHexError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        hex::decode(s).map(|v| Address::from_slice(v.as_slice()))
+    }
+}
+
+impl From<Address> for Hash {
```

Comment:
That looks weird, why do we need to convert between an Address and a Hash?

Reply:
If we currently do this in production code and removing that is non-trivial, can you please record a follow-up issue?

Reply:
Recorded as follow up: https://github.com/comit-network/comit-rs/issues/2411

The problem is that the [`Topic`](https://github.com/comit-network/comit-rs/blob/1ceaa33ee93ac80d465a90001d21a0c96079ce6a/cnd/src/btsieve/ethereum.rs#L230) used to search in `Event`s just accepts a `Hash` specifically, but we also use it for searching for `Address`es - we should change this to be generic.

Reply:
Okay interesting, I would put this functionality on the specific topic then and not as a generall conversion between addresses and hashes.